File: 05-binary_search/00074-Search_a_2D_Matrix/00074-search_a_2D_matrix.java
// 74: Search a 2D Matrix
// https://leetcode.com/problems/search-a-2d-matrix/


class Solution {
    // SOLUTION
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
		int cols = matrix[0].length;
        int row = 0;
        int col = cols-1;
			
        while (row<rows && col>-1) {
            int cur = matrix[row][col];
            if (cur == target) return true;
            if (target > cur) row++;
            else col--;
        }
        
        return false;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        int[][] matrix = {{1,3,5,7},{10,11,16,20},{23,30,34,60}};
        int target = 3;

        // OUTPUT
        var result = o.searchMatrix(matrix, target);
        System.out.println(result ? "true" : "false");
    }
}

File: 05-binary_search/00074-Search_a_2D_Matrix/00074-search_a_2D_matrix.py
# 74: Search a 2D Matrix
# https://leetcode.com/problems/search-a-2d-matrix/


class Solution:
    # SOLUTION
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        row = 0
        col = cols-1
			
        while row<rows and col>-1:
            cur = matrix[row][col]
            if cur == target: return True
            if target > cur: row+=1
            else: col-=1
        
        return False


if __name__ == "__main__":
    o = Solution()

    # INPUT
    matrix: list[list[int]] = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
    target: int = 3

    # OUTPUT
    result = o.searchMatrix(matrix, target)
    print(result)